packages/compass-shell/src/components/compass-shell/tab-compass-shell.tsx

```diff
@@ -93,26 +85,55 @@ const Shell = React.forwardRef<ShellType, ShellProps>(function Shell(
       shellRef.current?.focusEditor();
     });
   }, []);
+
+  return <_Shell ref={mergeRef} {...props}></_Shell>;
+});
+
+function useInitialEval(initialEvaluate?: string | string[]) {
+  const [initialEvalApplied, setInitialEvalApplied] = useTabState(
+    'initialEvalApplied',
+    false
+  );
+  useEffect(() => {
+    setInitialEvalApplied(true);
+  }, [setInitialEvalApplied]);
+  return initialEvalApplied ? undefined : initialEvaluate;
+}
+
+const normalizeInitialEvaluate = (initialEvaluate: string | string[]) => {
   return (
-    <_Shell
-      ref={mergeRef}
-      initialEvaluate={initialEvaluate}
-      {...props}
-    ></_Shell>
+    Array.isArray(initialEvaluate) ? initialEvaluate : [initialEvaluate]
+  ).filter((line) => {
+    // Filter out empty lines if passed by accident
+    return !!line;
+  });
+};
+
+const isInitialEvaluateEmpty = (
+  initialEvaluate?: string | string[] | undefined
+) => {
+  return (
+    !initialEvaluate || normalizeInitialEvaluate(initialEvaluate).length === 0
   );
-});
+};
 
 const CompassShell: React.FC<CompassShellProps> = ({
   runtime,
   initialHistory,
   onHistoryChange,
-  initialEvaluate,
+  initialEvaluate: _initialEvaluate,
   initialInput,
 }) => {
+  const initialEvaluate = useInitialEval(_initialEvaluate);
+
+  const [isOperationInProgress, setIsOperationInProgress] = useTabState(
```

Comment:
It would probably be nice to have this boolean even higher up so we can change the icon in the tab itself. That way you can see if the operation finishes while you're on a different tab, but that's probably a job for a different ticket.